### DISCUSSION TITLE
ReferenceError: window is not defined

### DISCUSSION THREAD

**Discussion by sushilbansal** [ceuk/speech-recognition-aws-polyfill#1] [Q&A]
I tried to use in the following way and got an error:

```
import React, { useEffect, useState } from 'react';
import SpeechRecognitionPolyfill from 'speech-recognition-aws-polyfill';

const recognition = new SpeechRecognitionPolyfill({
  IdentityPoolId: 'us-west-x:XXX-xxx-xxxxx-xxxxx', // My Identity Pool ID
  region: 'us-west-x', // My AWS region
  lang: 'en-GB'
});

export const AWSTranscribe = () => {
  const [results, setResults] = useState('');
  useEffect(() => {
    recognition.start();
  }, []);
  recognition.onresult = function (event: any) {
    const { transcript } = event.results[0][0];
    setResults(transcript);
    console.log('Heard: ', transcript);
  };

  recognition.onerror = (error: any) => {
    console.log('Somehting went wrong: --- ', error);
  };
  return <div>{results}</div>;
};
```

**Comment by sushilbansal** [comment]
@ceuk : probably i am not doing polyfill properly. Can you please suggest?

**Reply by ceuk** [reply]
It was my fault mate - upgrade to 0.1.7 and try again. Sorry about that.

**Reply by sushilbansal** [reply]
let me try that :) thanks for an update

**Reply by sushilbansal** [reply]
same error

**Reply by ceuk** [reply]
What browser are you using? And are you definitely using 0.1.7? I Just tried in Firefox and Chrome using your code in a create-react-app and it seemed to work ok

**Reply by sushilbansal** [reply]
1) using chrome on mac.
2) yes, it is 0.1.7
3) using nextjs
i am doing the below:
This TestTranscript is wrapped to make sure it is exucuted on client only:
```
if (window !== undefined) {
      return <TestTranscript />;
    }
    return <div />;
```
```
import React, { useState } from 'react';
import SpeechRecognitionPolyfill from 'speech-recognition-aws-polyfill';

export const TestTranscript = () => {
  const [results, setResults] = useState('');
  let recognition;
  if (window.SpeechRecognition) {
    recognition = new window.SpeechRecognition();
  } else if ((window as any).webkitSpeechRecognition) {
    recognition = new (window as any).webkitSpeechRecognition();
  } else {
    recognition = new SpeechRecognitionPolyfill({
      IdentityPoolId: 'us-east-2:xxxx-xxx-xxxx-xxx, // my Identity Pool ID
      region: 'us-east-2', // your AWS region
      lang: 'en-GB'
    });
  }
  recognition?.start();

  recognition.onresult = function (event) {
    var color = event.results[0][0].transcript;
    setResults((res) => res + ' ' + color);
    console.log(color);
  };
  return <div>{results}</div>;
};
```

**Reply by ceuk** [reply]
Ah, that probably explains it, there will be references to `window` inside the library before you even instantiate the class which is probably what is giving you the "window not defined" errors.

I'll refactor it so it plays nicely in environments without `window` and push the changes in 0.1.8.

For now, you could try using CommonJS imports instead so you can wrap the actual import inside an if statement

e.g. 

```javascript
let  SpeechRecognitionPolyfill
if (window) {
  SpeechRecognitionPolyfill = require('speech-recognition-aws-polyfill')
}
```

Super ugly but it should allow you to progress at least

**Reply by sushilbansal** [reply]
You are right; it works in CRA but not in nextjs. Testing out more on the same.

**Reply by ceuk** [reply]
Okay, give `0.1.8` a go - I think I've fixed all the unsafe references to `window` but they may still be some in my dependencies

**Reply by sushilbansal** [reply]
let me try it out.

**Reply by sushilbansal** [reply]
yeah the same window issue on latest repo. But i managed to work around that:
```
const DynamicComponent = dynamic(() => import('./AWSTranscribe'), {
  ssr: false
});

```

**Reply by ceuk** [reply]
I'll set up next.js this weekend and recreate/fix the issue

**Reply by sushilbansal** [reply]
Not sure if it is intended behavior: the audio to text works for 5 seconds and then it does not work. Tested both on create react app and next js in both dev and prod env. With standard SpeechRecognition it start coverting again after a pause. Does not happen in this case.

**Reply by ceuk** [reply]
It is intentional at the moment yeah until we support `continuous`
